blog/2025-08-22-pattern.mdx

```diff
@@ -0,0 +1,312 @@
+**tuples**, **vectors**, and **structures**. Understanding these types is fundamental
+to effectively using destructuring.
+
+- Tuple type `[T, ...]`: Tuple types are ordered, heterogeneous
+collections of types (e.g., `[Field, Boolean, Uint<8>]`).
+Destructuring tuple values allows you to bind individual
+elements to identifiers based on their position within a tuple value and typecheck
+the actual element type with the declared type annotation if one exists.
+- Vector type `Vector<n, T>`: Vector types are homogeneous tuple types
+(e.g., `Vector<3, Field>`).
+Destructuring vector values is the same as destructuring tuple values since
+vector values are constructed the same way tuple values are.
+- Structure type `S { f: T ... }`: Structure types are user-defined types with
+named fields (e.g., `struct Point { x: Field, y: Field }`).  Destructuring
+structure values lets you extract values of specific fields by their names and typecheck
+the actual field's type with its declared type annotation if one exists.
+
+## Where Can You Use It? 📍
+
+Pattern destructuring isn't just for a single use case; it's integrated into
+several key areas of Compact programming:
+
+- Parameters of circuits, anonymous circuits, and a constructor can be a pattern.
+- `const` binding statements can bind a pattern. This is arguably where you'll use destructuring
+most frequently for local variable assignments, allowing you to unpack values
+from expressions into new variables.
+
+## How to Destructure Values in Compact 🛠️
+
+Let's dive into the specifics of how to use tuple and structure patterns.
+The grammar for a pattern `p` is provided in
+[Compact's reference](https://docs.midnight.network/develop/reference/compact/lang-ref#parameters-patterns-and-destructuring).
+
+
+### Destructuring Tuple Values
+
+Tuple values can be destructured using a list of patterns enclosed in `[` and `]`.
+You can destructure a tuple into a sequence of identifiers:
+
+```compact
+const [a, b] = [1, true]; // 'a' is 1, 'b' is true
+```
+
+Sometimes you don't need all the values in a tuple. Compact allows you to skip
+elements by simply omitting the identifier but you cannot skip the actual element:
+
+```compact
+const [x, , z] = [10, 20, 30]; // 'x' is 10, 'z' is 30, 20 is skipped
+// const [x, , z] = [10, , 30]; // triggers a parse error
+```
+
+Skipped identifiers count towards the length of the pattern if there exists a
+named identifier after them. For example, `[x, y, ,]` has length 2 but
+`[x, , y]` has length 3.
+
+:::note
+If you want to enforce length 'l' of a tuple pattern you must have a named identifier
+at the `l-1` index of your pattern. For example, if you need to enforce a tuple
+pattern of legth 3 you need to have a named identifier on index 2 (e.g., all
+`[, , x]`, `[, x, y]`, and `[x, y, z]` enforce a tuple pattern of length 3).
+:::
+
+### Type Annotations for Tuple Patterns
+
+In a circuit definition, type declarations are required for tuple patterns:
```

Comment:
We haven't seen tuple patterns for parameters yet, it might be better to introduce them earlier.

I don't know if you want to spell it out, but it might be useful to give people the model that patterns for parameters are purely a convenience.  Without it they could get exactly the same effect by immediately (or just before they need to) destructuring the parameter:

```
circuit example(tuple0: [Field, Boolean]): Field {
  const [x, y] = tuple0;
  return x;
}
```

As a convenience we allow them to move the pattern into the circuit declaration by replacing the right-hand side identifier in the destructuring `tuple0` with the left-hand side pattern. (That also helps explain why the type is still needed there.)

Reply:
An alternative, which is a bigger rewrite, is to talk about *everything* in terms of const binding, and then only at the end talk about parameters as a special case of const binding.